File: algorithms/dynamic-programming/rodcutting.py
def rod_cutting_tb(n, p, r):
  if r[n] >= 0:
    return r[n]
  if n == 0:
    return 0
  ret = p[n]
  for i in range(1, n):
    ret = max(ret, rod_cutting_tb(i, p, r) + rod_cutting_tb(n-i, p, r))
  r[n] = ret
  return ret

def rod_cutting_bt(n, p):
  r = [-1 for i in range(len(p)+1)]
  r[0] = 0
  for j in range(1, n+1):
    q = p[j]
    for i in range(1, j):
      q = max(q, r[i] + r[j-i])
    r[j] = q 
  return r[n]

if __name__ == "__main__":
  p = [0, 1, 5, 8, 9, 10, 17, 17, 20, 24, 30]
  r = [-1 for i in range(len(p)+1)]
  print(rod_cutting_tb(10, p, r))
  print(rod_cutting_bt(10, p))

File: algorithms/dynamic-programming/rodcutting.c
#include <stdio.h>
#include <stdlib.h>

int rodCuttingTb(int n, int *p, int *r){
  if (r[n] >= 0)
    return r[n];
  if (n == 0)
    return 0;
  int ret = p[n];
  for (int i=1; i<n; i++) {
    int t = rodCuttingTb(i, p, r) + rodCuttingTb(n-i, p, r);
    if (ret < t) {
      ret = t;
    }
  }
  r[n] = ret;
  return ret;
}

int rodCuttingBt(int n, int *p, int p_n){
  int r[p_n];
  for (int i=1; i<p_n; i++) {
    r[i] = -1;
  }
  r[0] = 0;
  for (int j=1; j<n+1; j++) {
    int q = p[j];
    for (int i=1; i<j; i++) {
      int t = r[i] + r[j-i]; 
      if (q < t) {
        q = t;
      }
    }
    r[j] = q;
  }
  return r[n];
}

int main(){
  int p[] = {0, 1, 5, 8, 9, 10, 17, 17, 20, 24, 30};
  int p_n = sizeof(p)/sizeof(int);
  int r[p_n];
  for (int i=0; i<p_n; i++) {
    r[i] = -1;
  }
  printf("%d\n", rodCuttingTb(10, p, r));
  printf("%d\n", rodCuttingBt(10, p, p_n));
}
